File: paleo_workbench/ui/pages/map_edit_topology.py
from __future__ import annotations

from typing import Any

from paleo_workbench.mapping import map_edit_api as api
from paleo_workbench.mapping.geometry_schema import new_feature_id
from paleo_workbench.ui.pages.map_edit_commands import (
    BatchVertexEditCommand,
    CompositeCommand,
    CreateFeatureCommand,
    DeleteFeatureCommand,
)
from paleo_workbench.ui.pages.map_edit_items import FaciesPolygonItem, LineItem


def facies_geometry_issues(item: FaciesPolygonItem) -> list[dict[str, object]]:
    """Return geometry issues for a single facies polygon item."""
    issues: list[dict[str, object]] = []
    for part_index, ring_index, ring in item.iter_ring_addresses():
        for issue in api.validate_ring(ring):
            issues.append({
                "feature_id": item.feature_id,
                "part_index": part_index,
                "ring_index": ring_index,
                "code": str(issue.get("code", "invalid_geometry")),
                "message": str(issue.get("message", "几何无效")),
                "severity": "error",
            })
    try:
        from geoviz import validate_polygon_geometry

        shape_issues = validate_polygon_geometry(
            item.to_record()["geometry_type"], item.geometry_coordinates()
        )
    except (ImportError, AttributeError):
        shape_issues = []
    for issue in shape_issues:
        issues.append({
            "feature_id": item.feature_id,
            "code": str(issue.get("code", "invalid_shape")),
            "message": str(issue.get("message", "Shape 几何无效")),
            "severity": "error",
        })
    return issues


def apply_adjacency_warnings(items: list[FaciesPolygonItem], gap_tol: float) -> None:
    """Apply warning status to adjacent facies polygons with gaps."""
    if len(items) < 2:
        return
    rings = [item.coordinates() for item in items]
    issues = api.validate_adjacency(rings, gap_tol=gap_tol)
    flagged: set[int] = set()
    for issue in issues:
        pair = issue.get("pair")
        if isinstance(pair, (list, tuple)) and len(pair) == 2:
            flagged.add(int(pair[0]))
            flagged.add(int(pair[1]))
    for idx in flagged:
        if 0 <= idx < len(items) and items[idx].topology_status == "ok":
            items[idx].set_topology_status("warning")


def plan_topology_rebuild(
    facies_items: list[FaciesPolygonItem],
    tol: float,
    apply_coords_fn,
) -> tuple[dict[str, Any], BatchVertexEditCommand | None]:
    """Calculate forced topology rebuild report and command."""
    if not facies_items:
        return (
            {
                "snapped_count": 0,
                "ring_warnings": 0,
                "adjacency_issues": 0,
                "changed": False,
            },
            None,
        )
    rings = [item.coordinates() for item in facies_items]
    report = api.rebuild_topology(rings, snap_tol=tol, gap_tol=tol)
    snapped_rings = report["rings"]
    changes: list[tuple[str, list, list]] = []
    for item, old_ring, new_ring in zip(facies_items, rings, snapped_rings):
        if old_ring != new_ring:
            changes.append((item.feature_id, old_ring, new_ring))
    cmd = BatchVertexEditCommand(changes, apply_coordinates=apply_coords_fn) if changes else None
    res_report = {
        "snapped_count": len(changes),
        "ring_warnings": len(report.get("ring_issues") or []),
        "adjacency_issues": len(report.get("adjacency_issues") or []),
        "changed": bool(changes or report.get("changed")),
        "ring_issues": report.get("ring_issues") or [],
        "adjacency_issue_list": report.get("adjacency_issues") or [],
    }
    return res_report, cmd


def plan_merge_facies(
    a: FaciesPolygonItem,
    b: FaciesPolygonItem,
    add_feature_fn,
    remove_feature_fn,
    item_from_record_fn,
) -> tuple[str | None, CompositeCommand | None]:
    """Plan merger of two facies polygons into one composite command."""
    if a.has_complex_geometry() or b.has_complex_geometry():
        return None, None
    merged = api.merge_rings(a.coordinates(), b.coordinates())
    if not merged:
        return None, None
    name = (
        str(a.get_property("name") or "")
        or str(b.get_property("name") or "")
        or "合并相带"
    )
    new_id = new_feature_id("facies")
    style = dict(a.to_record().get("style") or {})
    new_rec = {
        "id": new_id,
        "kind": "facies",
        "name": name,
        "coordinates": merged,
        "style": style,
    }
    if item_from_record_fn(new_rec) is None:
        return None, None
    cmd = CompositeCommand([
        DeleteFeatureCommand(
            a.to_record(),
            add_feature=add_feature_fn,
            remove_feature=remove_feature_fn,
        ),
        DeleteFeatureCommand(
            b.to_record(),
            add_feature=add_feature_fn,
            remove_feature=remove_feature_fn,
        ),
        CreateFeatureCommand(
            new_rec,
            add_feature=add_feature_fn,
            remove_feature=remove_feature_fn,
        ),
    ])
    return new_id, cmd


def plan_split_facies(
    poly_item: FaciesPolygonItem,
    line_item: LineItem,
    add_feature_fn,
    remove_feature_fn,
    item_from_record_fn,
) -> tuple[list[str] | None, CompositeCommand | None]:
    """Plan splitting of one facies polygon by a line item."""
    if poly_item.has_complex_geometry():
        return None, None
    parts = api.split_ring_by_line(poly_item.coordinates(), line_item.coordinates())
    if not parts or len(parts) < 2:
        return None, None
    base_name = str(poly_item.get_property("name") or "") or "相带"
    poly_style = dict(poly_item.to_record().get("style") or {})
    new_recs = []
    new_ids: list[str] = []
    for i, ring in enumerate(parts):
        nid = new_feature_id("facies")
        new_ids.append(nid)
        new_recs.append({
            "id": nid,
            "kind": "facies",
            "name": f"{base_name}-{i + 1}",
            "coordinates": ring,
            "style": poly_style,
        })
    children: list = [
        DeleteFeatureCommand(
            poly_item.to_record(),
            add_feature=add_feature_fn,
            remove_feature=remove_feature_fn,
        )
    ]
    for rec in new_recs:
        if item_from_record_fn(rec) is None:
            return None, None
        children.append(
            CreateFeatureCommand(
                rec,
                add_feature=add_feature_fn,
                remove_feature=remove_feature_fn,
            )
        )
    return new_ids, CompositeCommand(children)

File: libs/ui_data_core/src/map_edit_topology.cpp
#include "pwb/ui_data_core/map_edit_topology.hpp"

#include "pwb/ui_data_core/json_util.hpp"

#include <set>
#include <utility>

namespace pwb::ui_data_core {
namespace {

domain::Json issue_entry(const FaciesPolygonModel& item,
                         const domain::Json& issue, bool with_address,
                         int part_index, int ring_index) {
    domain::Json entry = domain::Json::object();
    entry["feature_id"] = item.feature_id;
    if (with_address) {
        entry["part_index"] = part_index;
        entry["ring_index"] = ring_index;
    }
    entry["code"] = json_get_string(issue, "code",
                                    with_address ? "invalid_geometry"
                                                 : "invalid_shape");
    entry["message"] = json_get_string(issue, "message",
                                       with_address ? "几何无效"
                                                    : "Shape 几何无效");
    entry["severity"] = "error";
    return entry;
}

FeatureIdFn effective_ids(const FeatureIdFn& ids) {
    return ids ? ids
               : FeatureIdFn(
                     [](std::string_view p) { return new_feature_id(p); });
}

}  // namespace

domain::Json facies_geometry_issues(const FaciesPolygonModel& item,
                                    MapGeometryBackend* backend) {
    domain::Json issues = domain::Json::array();
    for (const auto& address : item.iter_ring_addresses()) {
        const domain::Json ring_issues = validate_ring(address.points);
        for (const auto& issue : ring_issues) {
            issues.push_back(issue_entry(item, issue, true,
                                         address.part_index,
                                         address.ring_index));
        }
    }
    // from geoviz import validate_polygon_geometry — ImportError /
    // AttributeError → [].
    const domain::Json shape_issues = validate_polygon_geometry(
        item.geometry_type, item.geometry_coordinates(), backend);
    for (const auto& issue : shape_issues) {
        issues.push_back(issue_entry(item, issue, false, 0, 0));
    }
    return issues;
}

void apply_adjacency_warnings(std::vector<FaciesPolygonModel*>& items,
                              double gap_tol) {
    if (items.size() < 2) {
        return;
    }
    std::vector<MapRing> rings;
    rings.reserve(items.size());
    for (const auto* item : items) {
        rings.push_back(item->coordinates());
    }
    const domain::Json issues = validate_adjacency(rings, gap_tol);
    std::set<int> flagged;
    for (const auto& issue : issues) {
        const auto it = issue.find("pair");
        if (it == issue.end() || !it->is_array() || it->size() != 2) {
            continue;
        }
        const auto a = (*it)[0].is_number() ? std::optional<int>((*it)[0].get<int>())
                                            : std::nullopt;
        const auto b = (*it)[1].is_number() ? std::optional<int>((*it)[1].get<int>())
                                            : std::nullopt;
        if (a.has_value()) {
            flagged.insert(*a);
        }
        if (b.has_value()) {
            flagged.insert(*b);
        }
    }
    for (const int idx : flagged) {
        if (idx >= 0 && idx < static_cast<int>(items.size()) &&
            items[idx]->topology_status == kTopologyOk) {
            items[idx]->set_topology_status(
                std::string_view{kTopologyWarning});
        }
    }
}

TopologyRebuildPlan plan_topology_rebuild(
    const std::vector<FaciesPolygonModel*>& facies_items, double tol,
    ApplyCoordinatesFn apply_coordinates) {
    TopologyRebuildPlan plan;
    if (facies_items.empty()) {
        plan.report["snapped_count"] = 0;
        plan.report["ring_warnings"] = 0;
        plan.report["adjacency_issues"] = 0;
        plan.report["changed"] = false;
        return plan;
    }
    std::vector<MapRing> rings;
    rings.reserve(facies_items.size());
    for (const auto* item : facies_items) {
        rings.push_back(item->coordinates());
    }
    const TopologyRebuildReport report =
        rebuild_topology(rings, tol, tol);
    std::vector<BatchVertexEditCommand::Change> changes;
    for (std::size_t i = 0; i < facies_items.size(); ++i) {
        if (rings[i] != report.rings[i]) {
            changes.push_back({facies_items[i]->feature_id, rings[i],
                               report.rings[i]});
        }
    }
    const long long snapped_count = static_cast<long long>(changes.size());
    if (!changes.empty()) {
        plan.command = std::make_shared<BatchVertexEditCommand>(
            std::move(changes), std::move(apply_coordinates));
    }
    plan.report["snapped_count"] = snapped_count;
    plan.report["ring_warnings"] =
        static_cast<long long>(report.ring_issues.size());
    plan.report["adjacency_issues"] =
        static_cast<long long>(report.adjacency_issues.size());
    plan.report["changed"] = snapped_count > 0 || report.changed;
    plan.report["ring_issues"] = report.ring_issues;
    plan.report["adjacency_issue_list"] = report.adjacency_issues;
    return plan;
}

MergeFaciesPlan plan_merge_facies(
    const FaciesPolygonModel& a, const FaciesPolygonModel& b,
    AddFeatureFn add_feature, RemoveFeatureFn remove_feature,
    const ItemFromRecordFn& item_from_record, MapGeometryBackend* backend,
    const FeatureIdFn& ids) {
    MergeFaciesPlan plan;
    if (a.has_complex_geometry() || b.has_complex_geometry()) {
        return plan;
    }
    const auto merged = merge_rings(a.coordinates(), b.coordinates(),
                                    backend);
    if (!merged.has_value()) {
        return plan;
    }
    // name = a.name or b.name or "合并相带"
    const domain::Json a_name = a.get_property("name");
    const domain::Json b_name = b.get_property("name");
    const std::string name =
        json_truthy(a_name)   ? python_str(a_name)
        : json_truthy(b_name) ? python_str(b_name)
                              : "合并相带";
    const FeatureIdFn id_gen = effective_ids(ids);
    const std::string new_id = id_gen("facies");
    // style = dict(a.to_record().get("style") or {})
    domain::Json style = domain::Json::object();
    {
        const domain::Json rec = a.to_record();
        const auto it = rec.find("style");
        if (it != rec.end() && it->is_object()) {
            style = *it;
        }
    }
    domain::Json new_rec = domain::Json::object();
    new_rec["id"] = new_id;
    new_rec["kind"] = "facies";
    new_rec["name"] = name;
    new_rec["coordinates"] = ring_to_json(*merged);
    new_rec["style"] = style;
    if (!item_from_record(new_rec).has_value()) {
        return plan;
    }
    auto delete_a = std::make_shared<DeleteFeatureCommand>(
        a.to_record(), add_feature, remove_feature);
    auto delete_b = std::make_shared<DeleteFeatureCommand>(
        b.to_record(), add_feature, remove_feature);
    auto create = std::make_shared<CreateFeatureCommand>(
        new_rec, add_feature, remove_feature);
    plan.command = std::make_shared<CompositeCommand>(
        std::vector<std::shared_ptr<EditCommand>>{delete_a, delete_b,
                                                  create});
    plan.new_id = new_id;
    return plan;
}

SplitFaciesPlan plan_split_facies(
    const FaciesPolygonModel& poly_item, const LineModel& line_item,
    AddFeatureFn add_feature, RemoveFeatureFn remove_feature,
    const ItemFromRecordFn& item_from_record, MapGeometryBackend* backend,
    const FeatureIdFn& ids) {
    SplitFaciesPlan plan;
    if (poly_item.has_complex_geometry()) {
        return plan;
    }
    const auto parts = split_ring_by_line(poly_item.coordinates(),
                                          line_item.coordinates(), backend);
    if (!parts.has_value() || parts->size() < 2) {
        return plan;
    }
    const domain::Json base = poly_item.get_property("name");
    const std::string base_name =
        json_truthy(base) ? python_str(base) : "相带";
    domain::Json poly_style = domain::Json::object();
    {
        const domain::Json rec = poly_item.to_record();
        const auto it = rec.find("style");
        if (it != rec.end() && it->is_object()) {
            poly_style = *it;
        }
    }
    const FeatureIdFn id_gen = effective_ids(ids);
    std::vector<domain::Json> new_recs;
    std::vector<std::string> new_ids;
    for (std::size_t i = 0; i < parts->size(); ++i) {
        const std::string nid = id_gen("facies");
        new_ids.push_back(nid);
        domain::Json rec = domain::Json::object();
        rec["id"] = nid;
        rec["kind"] = "facies";
        rec["name"] = base_name + "-" + std::to_string(i + 1);
        rec["coordinates"] = ring_to_json((*parts)[i]);
        rec["style"] = poly_style;
        new_recs.push_back(std::move(rec));
    }
    std::vector<std::shared_ptr<EditCommand>> children;
    children.push_back(std::make_shared<DeleteFeatureCommand>(
        poly_item.to_record(), add_feature, remove_feature));
    for (const auto& rec : new_recs) {
        if (!item_from_record(rec).has_value()) {
            return SplitFaciesPlan{};
        }
        children.push_back(std::make_shared<CreateFeatureCommand>(
            rec, add_feature, remove_feature));
    }
    plan.command = std::make_shared<CompositeCommand>(std::move(children));
    plan.new_ids = std::move(new_ids);
    return plan;
}

}  // namespace pwb::ui_data_core
